master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,470 @@
+package internal
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/google/uuid"
+	"github.com/labstack/echo/v4"
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/internal/api"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+)
+
+const (
+	// MIMEApplicationGZip is GZip's MIME type.
+	MIMEApplicationGZip = "application/gzip"
+	// MIMEApplicationZip is Zip's MIME type.
+	MIMEApplicationZip = "application/zip"
+)
+
+func storageConfig2Str(config any) string {
+	switch config.(type) {
+	case expconf.AzureConfig:
+		return "azure"
+	case expconf.GCSConfig:
+		return "gcs"
+	case expconf.HDFSConfig:
+		return "hdfs"
+	case expconf.S3Config:
+		return "s3"
+	case expconf.SharedFSConfig:
+		return "shared_fs"
+	default:
+		return "unknown"
+	}
+}
+
+type archiveWriter interface {
+	WriteHeader(path string, size int64) error
+	Write(b []byte) (int, error)
+	Close() error
+}
+
+type archiveClosers struct {
+	closers []io.Closer
+}
+
+// Close() closes all items in closers in reverse order.
+func (ac *archiveClosers) Close() error {
+	for i := len(ac.closers) - 1; i >= 0; i-- {
+		err := ac.closers[i].Close()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type tarArchiveWriter struct {
+	archiveClosers
+	tw *tar.Writer
+}
+
+func (aw *tarArchiveWriter) WriteHeader(path string, size int64) error {
+	hdr := tar.Header{
+		Name: path,
+		Mode: 0666,
+		Size: size,
+	}
+	if strings.HasSuffix(path, "/") {
+		// This a directory
+		hdr.Mode = 0777
+	}
+	return aw.tw.WriteHeader(&hdr)
+}
+
+func (aw *tarArchiveWriter) Write(p []byte) (int, error) {
+	return aw.tw.Write(p)
+}
+
+type zipArchiveWriter struct {
+	archiveClosers
+	zw        *zip.Writer
+	zwContent io.Writer
+}
+
+func (aw *zipArchiveWriter) WriteHeader(path string, size int64) error {
+	// Zip by default sets mode 0666 and 0777 for files and folders respectively
+	zwc, err := aw.zw.Create(path)
+	if err != nil {
+		return err
+	}
+	aw.zwContent = zwc
+	return nil
+}
+
+func (aw *zipArchiveWriter) Write(p []byte) (int, error) {
+	var w io.Writer
+	// Guard against the mistake where WriteHeader() is not called before
+	// calling Write(). The AWS SDK likely will not make this mistake but
+	// zipArchiveWriter is not just limited to being used with AWS.
+	if aw.zwContent == nil {
+		return 0, nil
+	}
+	w = aw.zwContent
+	return w.Write(p)
+}
+
+type delayWriter struct {
```

Comment:
I suppose NewWriterSize would meet the technical requirements, but it would also introduce an extra unnecessary data copy after the initial buffering period is over.  delayWriter only buffers the initial bytes and is transparent after that.

Reply:
Yeah, essentially the same except `delayWriter` no long buffers after the buffer is full once. Yeah, I think I can use `bufio`.